File: sdk/go/client_test.go
package xiaoguai_test

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	xiaoguai "github.com/xiaoguai-agent/xiaoguai-go-client"
)

// ---------------------------------------------------------------------------
// Test helpers
// ---------------------------------------------------------------------------

func newTestServer(t *testing.T, handler http.HandlerFunc) (*httptest.Server, *xiaoguai.Client) {
	t.Helper()
	srv := httptest.NewServer(handler)
	t.Cleanup(srv.Close)
	c, err := xiaoguai.NewClient(srv.URL, xiaoguai.WithToken("test-token"))
	require.NoError(t, err)
	return srv, c
}

func jsonHandler(t *testing.T, statusCode int, body interface{}) http.HandlerFunc {
	t.Helper()
	return func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(statusCode)
		_ = json.NewEncoder(w).Encode(body)
	}
}

// ---------------------------------------------------------------------------
// HotL — ListHotlPolicies
// ---------------------------------------------------------------------------

var samplePolicy = map[string]interface{}{
	"id":             "aaaa-bbbb-cccc-dddd",
	"tenant_id":      "tenant-1",
	"scope":          "llm_call",
	"window_seconds": 3600,
	"max_count":      100,
	"escalate_to":    "ops@example.com",
}

func TestListHotlPolicies_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, []interface{}{samplePolicy}))
	policies, err := c.ListHotlPolicies(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Len(t, policies, 1)
	assert.Equal(t, "aaaa-bbbb-cccc-dddd", policies[0].ID)
	assert.Equal(t, "llm_call", policies[0].Scope)
	assert.Equal(t, 3600, policies[0].WindowSeconds)
}

func TestListHotlPolicies_EmptyResult(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, []interface{}{}))
	policies, err := c.ListHotlPolicies(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Empty(t, policies)
}

func TestListHotlPolicies_WithScope(t *testing.T) {
	var capturedQuery string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		capturedQuery = r.URL.RawQuery
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(200)
		_ = json.NewEncoder(w).Encode([]interface{}{})
	}))
	t.Cleanup(srv.Close)
	c, err := xiaoguai.NewClient(srv.URL, xiaoguai.WithToken("tok"))
	require.NoError(t, err)

	_, err = c.ListHotlPolicies(context.Background(), "tenant-1", xiaoguai.WithScope("email"))
	require.NoError(t, err)
	assert.Contains(t, capturedQuery, "scope=email")
}

func TestListHotlPolicies_EmptyTenantID(t *testing.T) {
	c, err := xiaoguai.NewClient("http://localhost:9999")
	require.NoError(t, err)
	_, err = c.ListHotlPolicies(context.Background(), "")
	assert.ErrorContains(t, err, "tenantID")
}

func TestListHotlPolicies_404(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 404, map[string]string{"error": "not found"}))
	_, err := c.ListHotlPolicies(context.Background(), "tenant-1")
	require.Error(t, err)
	var nf *xiaoguai.NotFoundError
	assert.True(t, errors.As(err, &nf))
}

func TestListHotlPolicies_500Retries(t *testing.T) {
	calls := 0
	_, c := newTestServer(t, func(w http.ResponseWriter, _ *http.Request) {
		calls++
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(500)
		_ = json.NewEncoder(w).Encode(map[string]string{"error": "internal"})
	})
	_, err := c.ListHotlPolicies(context.Background(), "tenant-1")
	require.Error(t, err)
	var se *xiaoguai.ServerError
	assert.True(t, errors.As(err, &se))
	assert.Equal(t, 3, calls, "should retry 3 times")
}

// ---------------------------------------------------------------------------
// HotL — CreateHotlPolicy
// ---------------------------------------------------------------------------

func TestCreateHotlPolicy_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, samplePolicy))
	maxCount := 100
	policy, err := c.CreateHotlPolicy(context.Background(), xiaoguai.CreateHotlPolicyRequest{
		TenantID:      "tenant-1",
		Scope:         "llm_call",
		WindowSeconds: 3600,
		MaxCount:      &maxCount,
	})
	require.NoError(t, err)
	assert.Equal(t, "aaaa-bbbb-cccc-dddd", policy.ID)
}

func TestCreateHotlPolicy_ValidationError(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 400, map[string]string{"error": "max_count or max_usd required"}))
	_, err := c.CreateHotlPolicy(context.Background(), xiaoguai.CreateHotlPolicyRequest{
		TenantID:      "tenant-1",
		Scope:         "llm_call",
		WindowSeconds: 3600,
	})
	require.Error(t, err)
	var ve *xiaoguai.ValidationError
	assert.True(t, errors.As(err, &ve))
}

func TestCreateHotlPolicy_EmptyTenantID(t *testing.T) {
	c, err := xiaoguai.NewClient("http://localhost:9999")
	require.NoError(t, err)
	_, err = c.CreateHotlPolicy(context.Background(), xiaoguai.CreateHotlPolicyRequest{Scope: "x"})
	assert.ErrorContains(t, err, "TenantID")
}

// ---------------------------------------------------------------------------
// HotL — DeleteHotlPolicy
// ---------------------------------------------------------------------------

func TestDeleteHotlPolicy_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, map[string]string{"ok": "true"}))
	err := c.DeleteHotlPolicy(context.Background(), "aaaa-bbbb-cccc-dddd")
	require.NoError(t, err)
}

func TestDeleteHotlPolicy_NotFound(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 404, map[string]string{"error": "not found"}))
	err := c.DeleteHotlPolicy(context.Background(), "no-such-id")
	require.Error(t, err)
	var nf *xiaoguai.NotFoundError
	assert.True(t, errors.As(err, &nf))
}

// ---------------------------------------------------------------------------
// Outcomes — RecordOutcome
// ---------------------------------------------------------------------------

func TestRecordOutcome_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, map[string]bool{"ok": true}))
	ok, err := c.RecordOutcome(context.Background(), xiaoguai.RecordOutcomeRequest{
		TenantID:  "tenant-1",
		AgentName: "sales-bot",
		Kind:      "revenue_usd",
		Value:     1200.0,
	})
	require.NoError(t, err)
	assert.True(t, ok)
}

func TestRecordOutcome_ValidationError(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 422, map[string]string{"error": "invalid value"}))
	_, err := c.RecordOutcome(context.Background(), xiaoguai.RecordOutcomeRequest{
		TenantID:  "tenant-1",
		AgentName: "bot",
		Kind:      "revenue_usd",
		Value:     -1,
	})
	require.Error(t, err)
	var ve *xiaoguai.ValidationError
	assert.True(t, errors.As(err, &ve))
}

func TestRecordOutcome_EmptyKind(t *testing.T) {
	c, err := xiaoguai.NewClient("http://localhost:9999")
	require.NoError(t, err)
	_, err = c.RecordOutcome(context.Background(), xiaoguai.RecordOutcomeRequest{
		TenantID:  "tenant-1",
		AgentName: "bot",
	})
	assert.ErrorContains(t, err, "Kind")
}

// ---------------------------------------------------------------------------
// Outcomes — OutcomesSummary
// ---------------------------------------------------------------------------

var sampleSummaryResp = map[string]interface{}{
	"tenant_id": "tenant-1",
	"range":     "30d",
	"summary": map[string]interface{}{
		"by_kind": map[string]interface{}{
			"revenue_usd": map[string]interface{}{
				"count": 5,
				"sum":   6000.0,
				"avg":   1200.0,
			},
		},
	},
}

func TestOutcomesSummary_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, sampleSummaryResp))
	summary, err := c.OutcomesSummary(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Equal(t, "tenant-1", summary.TenantID)
	assert.Equal(t, "30d", summary.Range)
	bucket, ok := summary.ByKind["revenue_usd"]
	require.True(t, ok)
	assert.Equal(t, 5, bucket.Count)
	assert.InDelta(t, 6000.0, bucket.Sum, 0.01)
}

func TestOutcomesSummary_WithRange(t *testing.T) {
	var capturedQuery string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		capturedQuery = r.URL.RawQuery
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(200)
		_ = json.NewEncoder(w).Encode(sampleSummaryResp)
	}))
	t.Cleanup(srv.Close)
	c, err := xiaoguai.NewClient(srv.URL)
	require.NoError(t, err)

	_, err = c.OutcomesSummary(context.Background(), "tenant-1", xiaoguai.WithRange("7d"))
	require.NoError(t, err)
	assert.Contains(t, capturedQuery, "range=7d")
}

// ---------------------------------------------------------------------------
// Outcomes — OutcomesTimeseries
// ---------------------------------------------------------------------------

var sampleTimeseriesResp = map[string]interface{}{
	"tenant_id": "tenant-1",
	"range":     "7d",
	"days": []interface{}{
		map[string]interface{}{"date": "2026-05-25", "kind": "revenue_usd", "count": 2, "sum": 2400.0},
	},
}

func TestOutcomesTimeseries_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, sampleTimeseriesResp))
	ts, err := c.OutcomesTimeseries(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Len(t, ts.Days, 1)
	assert.Equal(t, "2026-05-25", ts.Days[0].Date)
	assert.InDelta(t, 2400.0, ts.Days[0].Sum, 0.01)
}

func TestOutcomesTimeseries_EmptyDays(t *testing.T) {
	resp := map[string]interface{}{"tenant_id": "tenant-1", "range": "7d", "days": []interface{}{}}
	_, c := newTestServer(t, jsonHandler(t, 200, resp))
	ts, err := c.OutcomesTimeseries(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Empty(t, ts.Days)
}

// ---------------------------------------------------------------------------
// Skills — ListSkillCatalog
// ---------------------------------------------------------------------------

var sampleCatalogResp = map[string]interface{}{
	"version": 1,
	"packs": []interface{}{
		map[string]interface{}{
			"slug": "rag-legal", "name": "Legal RAG", "description": "Legal document QA",
			"version": "1.0.0", "category": "rag", "requires": map[string]interface{}{},
			"knobs": map[string]interface{}{}, "screenshot_url": nil,
		},
	},
}

func TestListSkillCatalog_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, sampleCatalogResp))
	packs, err := c.ListSkillCatalog(context.Background())
	require.NoError(t, err)
	assert.Len(t, packs, 1)
	assert.Equal(t, "rag-legal", packs[0].Slug)
	assert.Equal(t, "rag", packs[0].Category)
}

// ---------------------------------------------------------------------------
// Skills — ListInstalledSkills
// ---------------------------------------------------------------------------

var sampleInstalledPack = map[string]interface{}{
	"id": "inst-1", "tenant_id": "tenant-1", "pack_slug": "rag-legal",
	"version": "1.0.0", "config": map[string]interface{}{"top_k": 5},
	"installed_at": "2026-05-25T00:00:00Z",
}

func TestListInstalledSkills_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, []interface{}{sampleInstalledPack}))
	packs, err := c.ListInstalledSkills(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Len(t, packs, 1)
	assert.Equal(t, "rag-legal", packs[0].PackSlug)
}

func TestListInstalledSkills_Empty(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, []interface{}{}))
	packs, err := c.ListInstalledSkills(context.Background(), "tenant-x")
	require.NoError(t, err)
	assert.Empty(t, packs)
}

// ---------------------------------------------------------------------------
// Skills — InstallSkill
// ---------------------------------------------------------------------------

func TestInstallSkill_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, sampleInstalledPack))
	pack, err := c.InstallSkill(context.Background(), xiaoguai.InstallSkillRequest{
		TenantID: "tenant-1",
		PackSlug: "rag-legal",
	})
	require.NoError(t, err)
	assert.Equal(t, "rag-legal", pack.PackSlug)
}

func TestInstallSkill_NotFound(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 404, map[string]string{"error": "not found"}))
	_, err := c.InstallSkill(context.Background(), xiaoguai.InstallSkillRequest{
		TenantID: "tenant-1",
		PackSlug: "unknown-slug",
	})
	require.Error(t, err)
	var nf *xiaoguai.NotFoundError
	assert.True(t, errors.As(err, &nf))
}

func TestInstallSkill_AlreadyInstalled(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 409, map[string]string{"error": "pack already installed"}))
	_, err := c.InstallSkill(context.Background(), xiaoguai.InstallSkillRequest{
		TenantID: "tenant-1",
		PackSlug: "rag-legal",
	})
	require.Error(t, err)
	var ce *xiaoguai.ConflictError
	assert.True(t, errors.As(err, &ce))
}

// ---------------------------------------------------------------------------
// Skills — UninstallSkill
// ---------------------------------------------------------------------------

func TestUninstallSkill_HappyPath(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 200, map[string]string{"deleted": "inst-1"}))
	deleted, err := c.UninstallSkill(context.Background(), "inst-1")
	require.NoError(t, err)
	assert.Equal(t, "inst-1", deleted)
}

func TestUninstallSkill_NotFound(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 404, map[string]string{"error": "not found"}))
	_, err := c.UninstallSkill(context.Background(), "does-not-exist")
	require.Error(t, err)
	var nf *xiaoguai.NotFoundError
	assert.True(t, errors.As(err, &nf))
}

// ---------------------------------------------------------------------------
// Error type hierarchy
// ---------------------------------------------------------------------------

func TestErrorHierarchy(t *testing.T) {
	cases := []struct {
		status int
		target interface{ Unwrap() error }
	}{
		{401, &xiaoguai.AuthError{}},
		{404, &xiaoguai.NotFoundError{}},
		{400, &xiaoguai.ValidationError{}},
		{409, &xiaoguai.ConflictError{}},
		{429, &xiaoguai.RateLimitError{}},
	}
	for _, tc := range cases {
		_, c := newTestServer(t, jsonHandler(t, tc.status, map[string]string{"error": "test"}))
		_, err := c.ListHotlPolicies(context.Background(), "t")
		require.Error(t, err)
		var base *xiaoguai.HTTPError
		assert.True(t, errors.As(err, &base), "expected HTTPError base for %d", tc.status)
	}
}

func TestGeneric5xxRaisesServerError(t *testing.T) {
	_, c := newTestServer(t, jsonHandler(t, 500, map[string]string{"error": "internal"}))
	_, err := c.ListHotlPolicies(context.Background(), "tenant-1")
	require.Error(t, err)
	var se *xiaoguai.ServerError
	assert.True(t, errors.As(err, &se))
	assert.Equal(t, 500, se.StatusCode)
}

// ---------------------------------------------------------------------------
// Context cancellation
// ---------------------------------------------------------------------------

func TestContextCancel(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(500 * time.Millisecond)
		w.WriteHeader(200)
		_, _ = w.Write([]byte("[]"))
	}))
	t.Cleanup(srv.Close)

	c, err := xiaoguai.NewClient(srv.URL)
	require.NoError(t, err)

	ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
	defer cancel()

	_, err = c.ListHotlPolicies(ctx, "tenant-1")
	require.Error(t, err)
}

// ---------------------------------------------------------------------------
// Custom HTTP client injection
// ---------------------------------------------------------------------------

func TestCustomHTTPClient(t *testing.T) {
	srv := httptest.NewServer(jsonHandler(t, 200, []interface{}{samplePolicy}))
	t.Cleanup(srv.Close)

	custom := &http.Client{Timeout: 5 * time.Second}
	c, err := xiaoguai.NewClient(srv.URL, xiaoguai.WithHTTPClient(custom))
	require.NoError(t, err)

	policies, err := c.ListHotlPolicies(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Len(t, policies, 1)
}

// ---------------------------------------------------------------------------
// Auth header
// ---------------------------------------------------------------------------

func TestAuthHeaderIsSent(t *testing.T) {
	var capturedAuth string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		capturedAuth = r.Header.Get("Authorization")
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(200)
		_, _ = w.Write([]byte("[]"))
	}))
	t.Cleanup(srv.Close)

	c, err := xiaoguai.NewClient(srv.URL, xiaoguai.WithToken("my-secret-token"))
	require.NoError(t, err)
	_, err = c.ListHotlPolicies(context.Background(), "tenant-1")
	require.NoError(t, err)
	assert.Equal(t, "Bearer my-secret-token", capturedAuth)
}

// ---------------------------------------------------------------------------
// NewClient validation
// ---------------------------------------------------------------------------

func TestNewClient_EmptyBaseURL(t *testing.T) {
	_, err := xiaoguai.NewClient("")
	require.Error(t, err)
}

// ---------------------------------------------------------------------------
// HotlVerdict helpers
// ---------------------------------------------------------------------------

func TestHotlVerdictAllowed(t *testing.T) {
	v := xiaoguai.HotlVerdict{Verdict: "allow"}
	assert.True(t, v.Allowed())
	assert.False(t, v.Denied())
}

func TestHotlVerdictDenied(t *testing.T) {
	reason := "budget exceeded"
	v := xiaoguai.HotlVerdict{Verdict: "deny", Reason: &reason}
	assert.True(t, v.Denied())
	assert.False(t, v.Allowed())
}

File: sdk/typescript/test/client.test.ts
/**
 * XiaoguaiClient test suite — 25 vitest cases covering:
 * - Happy-path for all implemented endpoints
 * - 4xx error mapping (401, 403, 404, 409, 422, 429)
 * - 5xx error mapping
 * - AbortSignal cancellation
 * - Custom fetch injection
 * - Retry-After header parsing
 * - NotImplementedError for stub methods
 * - Query parameter serialization
 */

import { describe, it, expect, vi } from "vitest";
import {
  XiaoguaiClient,
  XiaoguaiError,
  HttpError,
  AuthError,
  ForbiddenError,
  NotFoundError,
  ConflictError,
  ValidationError,
  RateLimitError,
  ServerError,
  NotImplementedError,
} from "../src/index.js";
import type { HotlPolicy, InstalledSkillPack } from "../src/index.js";

// ---------------------------------------------------------------------------
// Test helpers
// ---------------------------------------------------------------------------

/** Build a mock fetch that returns a fixed response. */
function mockFetch(status: number, body: unknown, headers: Record<string, string> = {}): typeof fetch {
  return vi.fn(async (_url: RequestInfo | URL, _init?: RequestInit) => {
    const responseHeaders = new Headers({
      "Content-Type": "application/json",
      ...headers,
    });
    return new Response(JSON.stringify(body), {
      status,
      headers: responseHeaders,
    });
  }) as unknown as typeof fetch;
}

/** Capture the URL that was requested. */
function capturingFetch(
  status: number,
  body: unknown,
): { fetch: typeof fetch; lastUrl: () => string } {
  let last = "";
  const fn = vi.fn(async (url: RequestInfo | URL, _init?: RequestInit) => {
    last = String(url);
    return new Response(JSON.stringify(body), {
      status,
      headers: new Headers({ "Content-Type": "application/json" }),
    });
  }) as unknown as typeof fetch;
  return { fetch: fn, lastUrl: () => last };
}

const BASE = "http://localhost:8080";

function client(fetchImpl: typeof fetch): XiaoguaiClient {
  return new XiaoguaiClient({ baseUrl: BASE, token: "test-token", fetch: fetchImpl });
}

// ---------------------------------------------------------------------------
// HotL — happy-path tests
// ---------------------------------------------------------------------------

describe("HotL policies", () => {
  const policyFixture: HotlPolicy = {
    id: "aaaabbbb-0000-0000-0000-000000000001",
    tenant_id: "tenant-1",
    scope: "llm_call",
    window_seconds: 3600,
    max_count: 100,
    max_usd: null,
    escalate_to: "ops@example.com",
  };

  it("listHotlPolicies — returns array on 200", async () => {
    const c = client(mockFetch(200, [policyFixture]));
    const result = await c.listHotlPolicies({ tenant_id: "tenant-1" });
    expect(result).toHaveLength(1);
    expect(result[0]!.scope).toBe("llm_call");
    expect(result[0]!.max_count).toBe(100);
  });

  it("listHotlPolicies — sends scope param in query string", async () => {
    const { fetch: fn, lastUrl } = capturingFetch(200, []);
    const c = client(fn);
    await c.listHotlPolicies({ tenant_id: "tenant-1", scope: "email_send" });
    expect(lastUrl()).toContain("scope=email_send");
    expect(lastUrl()).toContain("tenant_id=tenant-1");
  });

  it("createHotlPolicy — sends POST and returns policy", async () => {
    const fn = mockFetch(201, policyFixture);
    const c = client(fn);
    const result = await c.createHotlPolicy({
      tenant_id: "tenant-1",
      scope: "llm_call",
      window_seconds: 3600,
      max_count: 100,
    });
    expect(result.id).toBe(policyFixture.id);
    const calls = (fn as ReturnType<typeof vi.fn>).mock.calls;
    expect(calls[0]![1]!.method).toBe("POST");
  });

  it("deleteHotlPolicy — resolves void on 204", async () => {
    const c = client(
      vi.fn(async () =>
        new Response(null, { status: 204 }),
      ) as unknown as typeof fetch,
    );
    await expect(c.deleteHotlPolicy("some-id")).resolves.toBeUndefined();
  });

  it("getHotlPolicy — throws NotImplementedError", () => {
    const c = client(mockFetch(200, {}));
    expect(() => c.getHotlPolicy("any-id")).toThrow(NotImplementedError);
  });

  it("updateHotlPolicy — throws NotImplementedError", () => {
    const c = client(mockFetch(200, {}));
    expect(() => c.updateHotlPolicy("any-id", {})).toThrow(NotImplementedError);
  });

  it("checkHotl — throws NotImplementedError", () => {
    const c = client(mockFetch(200, {}));
    expect(() => c.checkHotl("llm_call", 1.0)).toThrow(NotImplementedError);
  });
});

// ---------------------------------------------------------------------------
// Outcomes — happy-path tests
// ---------------------------------------------------------------------------

describe("Outcomes", () => {
  it("recordOutcome — returns true on 201 ok:true", async () => {
    const c = client(mockFetch(201, { ok: true }));
    const ok = await c.recordOutcome({
      tenant_id: "t1",
      agent_name: "sales-bot",
      kind: "revenue_usd",
      value: 500,
    });
    expect(ok).toBe(true);
  });

  it("outcomesSummary — returns parsed summary", async () => {
    const fixture = {
      tenant_id: "t1",
      range: "30d",
      summary: { by_kind: { revenue_usd: { count: 3, sum: 1500, avg: 500 } } },
    };
    const c = client(mockFetch(200, fixture));
    const result = await c.outcomesSummary({ tenant_id: "t1", range: "30d" });
    expect(result.summary.by_kind["revenue_usd"]!.sum).toBe(1500);
  });

  it("outcomesTimeseries — returns days array", async () => {
    const fixture = {
      tenant_id: "t1",
      range: "7d",
      days: [{ date: "2026-05-25", kind: "revenue_usd", count: 2, sum: 800 }],
    };
    const c = client(mockFetch(200, fixture));
    const result = await c.outcomesTimeseries({ tenant_id: "t1", range: "7d" });
    expect(result.days).toHaveLength(1);
    expect(result.days[0]!.sum).toBe(800);
  });

  it("listOutcomes — throws NotImplementedError", () => {
    const c = client(mockFetch(200, []));
    expect(() => c.listOutcomes()).toThrow(NotImplementedError);
  });
});

// ---------------------------------------------------------------------------
// Skills — happy-path tests
// ---------------------------------------------------------------------------

describe("Skills", () => {
  const packFixture: InstalledSkillPack = {
    id: "row-id-1",
    tenant_id: "t1",
    pack_slug: "rag-legal",
    version: "1.0.0",
    config: {},
    installed_at: "2026-05-25T00:00:00Z",
  };

  it("listInstalledSkills — returns array", async () => {
    const c = client(mockFetch(200, [packFixture]));
    const result = await c.listInstalledSkills("t1");
    expect(result[0]!.pack_slug).toBe("rag-legal");
  });

  it("listInstalledSkills — passes tenant query param", async () => {
    const { fetch: fn, lastUrl } = capturingFetch(200, []);
    const c = client(fn);
    await c.listInstalledSkills("my-tenant");
    expect(lastUrl()).toContain("tenant=my-tenant");
  });

  it("listSkillCatalog — extracts packs from catalog response", async () => {
    const catalog = {
      version: 1,
      packs: [{ slug: "rag-hr", name: "RAG HR", description: "...", version: "1.0.0", category: "hr" }],
    };
    const c = client(mockFetch(200, catalog));
    const packs = await c.listSkillCatalog();
    expect(packs).toHaveLength(1);
    expect(packs[0]!.slug).toBe("rag-hr");
  });

  it("installSkill — sends POST, returns installed pack", async () => {
    const fn = mockFetch(200, packFixture);
    const c = client(fn);
    const result = await c.installSkill({ tenant_id: "t1", pack_slug: "rag-legal" });
    expect(result.id).toBe("row-id-1");
    expect((fn as ReturnType<typeof vi.fn>).mock.calls[0]![1]!.method).toBe("POST");
  });

  it("uninstallSkill — returns deleted id", async () => {
    const c = client(mockFetch(200, { deleted: "row-id-1" }));
    const deleted = await c.uninstallSkill("row-id-1");
    expect(deleted).toBe("row-id-1");
  });
});

// ---------------------------------------------------------------------------
// Error mapping tests
// ---------------------------------------------------------------------------

describe("Error mapping", () => {
  it("401 → AuthError", async () => {
    const c = client(mockFetch(401, { error: "Unauthorized" }));
    await expect(c.listHotlPolicies({ tenant_id: "t1" })).rejects.toThrow(AuthError);
  });

  it("401 → is instance of HttpError and XiaoguaiError", async () => {
    const c = client(mockFetch(401, {}));
    const err = await c.listHotlPolicies({ tenant_id: "t1" }).catch((e) => e);
    expect(err).toBeInstanceOf(HttpError);
    expect(err).toBeInstanceOf(XiaoguaiError);
    expect(err.status).toBe(401);
  });

  it("403 → ForbiddenError", async () => {
    const c = client(mockFetch(403, { error: "Forbidden" }));
    await expect(c.listSkillCatalog()).rejects.toThrow(ForbiddenError);
  });

  it("404 → NotFoundError", async () => {
    const c = client(mockFetch(404, { error: "not found" }));
    await expect(c.deleteHotlPolicy("bad-id")).rejects.toThrow(NotFoundError);
  });

  it("409 → ConflictError", async () => {
    const c = client(mockFetch(409, { error: "pack already installed" }));
    await expect(
      c.installSkill({ tenant_id: "t1", pack_slug: "rag-hr" }),
    ).rejects.toThrow(ConflictError);
  });

  it("422 → ValidationError", async () => {
    const c = client(mockFetch(422, { error: "invalid request" }));
    await expect(
      c.createHotlPolicy({ tenant_id: "t1", scope: "llm", window_seconds: -1 }),
    ).rejects.toThrow(ValidationError);
  });

  it("400 → ValidationError", async () => {
    const c = client(mockFetch(400, { error: "bad request" }));
    await expect(
      c.recordOutcome({ tenant_id: "", agent_name: "", kind: "", value: -1 }),
    ).rejects.toThrow(ValidationError);
  });

  it("429 → RateLimitError with retryAfter", async () => {
    const c = client(mockFetch(429, { error: "rate limited" }, { "retry-after": "60" }));
    const err = await c.listHotlPolicies({ tenant_id: "t1" }).catch((e) => e);
    expect(err).toBeInstanceOf(RateLimitError);
    expect(err.retryAfter).toBe(60);
  });

  it("500 → ServerError", async () => {
    const c = client(mockFetch(500, { error: "internal server error" }));
    await expect(c.listHotlPolicies({ tenant_id: "t1" })).rejects.toThrow(ServerError);
  });

  it("503 → ServerError", async () => {
    const c = client(mockFetch(503, { error: "service unavailable" }));
    await expect(c.outcomesSummary({ tenant_id: "t1" })).rejects.toThrow(ServerError);
  });
});

// ---------------------------------------------------------------------------
// AbortSignal cancellation
// ---------------------------------------------------------------------------

describe("AbortSignal", () => {
  it("aborts in-flight request when signal fires", async () => {
    const controller = new AbortController();
    let fetchCalled = false;
    const slowFetch = vi.fn(async (_url: RequestInfo | URL, init?: RequestInit) => {
      fetchCalled = true;
      // Wait until the signal aborts.
      return new Promise<Response>((_resolve, reject) => {
        if (init?.signal?.aborted) {
          reject(new DOMException("aborted", "AbortError"));
          return;
        }
        const onAbort = () => reject(new DOMException("aborted", "AbortError"));
        init?.signal?.addEventListener("abort", onAbort, { once: true });
      });
    }) as unknown as typeof fetch;

    const c = new XiaoguaiClient({ baseUrl: BASE, token: "tok", fetch: slowFetch });
    controller.abort();

    await expect(
      c.listHotlPolicies({ tenant_id: "t1" }, controller.signal),
    ).rejects.toThrow();
    expect(fetchCalled).toBe(true);
  });

  it("pre-aborted signal propagates to fetch call", async () => {
    const controller = new AbortController();
    controller.abort();
    // Simulate a fetch that honours a pre-aborted signal by rejecting.
    const abortAwareFetch = vi.fn(async (_url: RequestInfo | URL, init?: RequestInit) => {
      if (init?.signal?.aborted) {
        throw new DOMException("signal aborted", "AbortError");
      }
      return new Response(JSON.stringify([]), {
        status: 200,
        headers: new Headers({ "Content-Type": "application/json" }),
      });
    }) as unknown as typeof fetch;

    const c = new XiaoguaiClient({ baseUrl: BASE, token: "tok", fetch: abortAwareFetch });
    await expect(
      c.listHotlPolicies({ tenant_id: "t1" }, controller.signal),
    ).rejects.toThrow();
  });
});

// ---------------------------------------------------------------------------
// Custom fetch injection
// ---------------------------------------------------------------------------

describe("Custom fetch", () => {
  it("uses the provided fetch function instead of global", async () => {
    let called = false;
    const customFetch = vi.fn(async () => {
      called = true;
      return new Response(JSON.stringify([]), {
        status: 200,
        headers: new Headers({ "Content-Type": "application/json" }),
      });
    }) as unknown as typeof fetch;

    const c = new XiaoguaiClient({ baseUrl: BASE, fetch: customFetch });
    await c.listSkillCatalog();
    expect(called).toBe(true);
  });

  it("sends Authorization header when token is provided", async () => {
    let capturedHeaders: HeadersInit | undefined;
    const fn = vi.fn(async (_url: RequestInfo | URL, init?: RequestInit) => {
      capturedHeaders = init?.headers;
      return new Response(JSON.stringify({ version: 1, packs: [] }), {
        status: 200,
        headers: new Headers({ "Content-Type": "application/json" }),
      });
    }) as unknown as typeof fetch;

    const c = new XiaoguaiClient({ baseUrl: BASE, token: "secret-token", fetch: fn });
    await c.listSkillCatalog();
    const headers = capturedHeaders as Record<string, string>;
    expect(headers["Authorization"]).toBe("Bearer secret-token");
  });

  it("omits Authorization header when no token", async () => {
    let capturedHeaders: HeadersInit | undefined;
    const fn = vi.fn(async (_url: RequestInfo | URL, init?: RequestInit) => {
      capturedHeaders = init?.headers;
      return new Response(JSON.stringify({ version: 1, packs: [] }), {
        status: 200,
        headers: new Headers({ "Content-Type": "application/json" }),
      });
    }) as unknown as typeof fetch;

    const c = new XiaoguaiClient({ baseUrl: BASE, fetch: fn });
    await c.listSkillCatalog();
    const headers = capturedHeaders as Record<string, string>;
    expect(headers["Authorization"]).toBeUndefined();
  });
});
